quint/src/effects/inferrer.ts

```diff
@@ -149,23 +149,13 @@ export class EffectInferrer implements IRVisitor {
     }
 
     this.location = `Trying to infer effect for operator application in ${expressionToString(expr)}`
-    const paramsResult = mergeInMany(
-      expr.args.map((a: QuintEx) => {
-        return this.fetchResult(a.id).map(e => this.newInstance(e))
-      })
-    )
 
-    const resultEffect: Effect = { kind: 'variable', name: this.freshVarGenerator.freshVar('_e') }
-    const arrowEffect = paramsResult
-      .map(params => {
-        const effect: Effect = {
-          kind: 'arrow',
-          params,
-          result: resultEffect,
-        }
+    // Fetch all the effects for the arguments
+    const argEffects = mergeInMany(expr.args.map((a: QuintEx) => this.fetchResult(a.id).map(e => e.effect)))
```

Comment:
I feel like we do need the `newInstance` here. I'm thinking about a scenario like:

```
def f(x) = x // polymorphic operator
def op(f1: (a) => a, f2: (b) => b, x1: a, x2: b): (a,b) = (f1(x1), f2(x2))
val foo = op(f, f, 1, "a")
```

When fetching the results for the arguments, the first and the second argument (`f` and `f`) need to have different types (i.e. `(t0) => t0` and `(t1) => t1`), while with your approach, they would have the same types and I believe typechecking would fail.

Reply:
ok, this is actually for the effects and not types, but afaiu it's all the same. Perhaps we can't reproduce this issue in the effect system right now, tho. But I still feel like the correct thing here would be to use `newInstance`

Reply:
Ah, I see. I can revert this part then.